daffodil-io/src/main/scala/org/apache/daffodil/io/LimitingJavaIOStreams.scala

```diff
@@ -0,0 +1,281 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.io
+
+import org.apache.daffodil.exceptions.Assert
+import java.util.regex.Pattern
+import java.util.Scanner
+import java.nio.charset.StandardCharsets
+import java.nio.charset.Charset
+import java.io.InputStream
+import java.io.FilterInputStream
+import java.io.InputStreamReader
+
+/**
+ * This class can be used with any InputStream to restrict what is
+ * read from it to N bytes.
+ *
+ * This can be used to forcibly stop consumption of data from a stream at
+ * a length obtained explicitly.
+ *
+ * Thread safety: This is inherently stateful - so not thread safe to use
+ * this object from more than one thread.
+ */
+class ExplicitLengthLimitingStream(in: InputStream, limit: Int)
+  extends FilterInputStream(in) {
+
+  private var numRemaining = limit
+
+  override def read(buf: Array[Byte], off: Int, len: Int) = {
+    Assert.invariant(numRemaining >= 0)
+    if (numRemaining == 0) -1
+    else if (len == 0) 0
+    else {
+      val requestSize = math.min(numRemaining, len)
+      val actualSize = in.read(buf, off, requestSize)
+      if (actualSize == -1)
+        numRemaining = 0
+      else
+        numRemaining -= actualSize
+      actualSize
+    }
+  }
+
+  private val readBuf = new Array[Byte](1)
+
+  override def read(): Int = {
+    readBuf(0) = 0
+    val n = read(readBuf, 0, 1)
+    if (n == -1)
+      -1
+    else {
+      Assert.invariant(n == 1)
+      val i = readBuf(0).toInt
+      val b = i & 0xFF
+      b
+    }
+  }
+}
+
+/**
+ * This class can be used with any InputStream to restrict what is
+ * read from it to stop before a terminator string.
+ *
+ * The terminator string is exactly that, a string of characters. Not a
+ * regex, nor anything involving DFDL Character Entities or Character Class
+ * Entities. (No %WSP; no %NL; )
+ *
+ * This can be used to forcibly stop consumption of data from a stream at
+ * a length obtained from a delimiter.
+ *
+ * The terminator string is consumed from the underlying stream (if found), and
+ * the underlying stream is left positioned at the byte after the terminator
+ * string.
+ *
+ * Thread safety: This is inherently stateful - so not thread safe to use
+ * this object from more than one thread.
+ */
+class DelimitedLengthLimitingStream(inWithMarkSupported: InputStream, terminator: String,
+  charset: Charset = StandardCharsets.ISO_8859_1,
+  maxChunkSize: Int = 1024 * 32) // 32K chunk size by default. Units are chars.
+  extends FilterInputStream(inWithMarkSupported) {
```

Comment:
I think each layer DataInputStream, or Java.io.InputStream has to support mark/reset. When we wrap a JavaIOInputStream around a DataInputStream, we implement the mark/reset in terms of what DataInputStream supports. 

The need for mark/reset here is for a very knarly painful thing about the way I'm using Java Scanner class to search for a regex. Because this operats on a reader, and both it and the reader layer read ahead a bit, we have to play tricks to get the underlying input stream to be precisely positioned immediately after the last byte that was "required" to decode the layer. 

More in next commit.